File: src/javascript/algorithms/hashtable/8.letterCombinations/letterCombinations.js
function letterCombinations(digits){
    let combinations = [];
    if(digits.length === 0) {
        return combinations;
    }
    let digitToChar = new Map();

    digitToChar.set('2', "abc");
    digitToChar.set('3', "def");
    digitToChar.set('4', "ghi");
    digitToChar.set('5', "jkl");
    digitToChar.set('6', "mno");
    digitToChar.set('7', "pqrs");
    digitToChar.set('8', "tuv");
    digitToChar.set('9', "wxyz");

    backtrack(0, "", digits, digitToChar, combinations);

    return combinations;
}

function backtrack(i, currStr, digits, digitToChar, combinations){
    if(digits.length === currStr.length) {
        combinations.push(currStr);
        return;
    }

    for (const ch of digitToChar.get(digits[i])) {
        backtrack(i+1, currStr+ch, digits, digitToChar, combinations);
    }
}

let digits1 = "24";
let digits2 = "";
console.log(letterCombinations(digits1));
console.log(letterCombinations(digits2));

File: src/java1/algorithms/hashmap/letterCombinations/LetterCombinations.java
package java1.algorithms.hashmap.letterCombinations;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class LetterCombinations {
        private static final Map<Character, String> phoneMap = new HashMap<>();

        static {
            phoneMap.put('2', "abc");
            phoneMap.put('3', "def");
            phoneMap.put('4', "ghi");
            phoneMap.put('5', "jkl");
            phoneMap.put('6', "mno");
            phoneMap.put('7', "pqrs");
            phoneMap.put('8', "tuv");
            phoneMap.put('9', "wxyz");
        }


    //TC: O(n 4^n) SC: O(n 4^n)
    private static List<String> letterCombinations(String digits){
        List<String> combinations = new ArrayList<>();
        if(digits == null || digits.isEmpty()) {
            return combinations;
        }

        backtrack(0, new StringBuilder(), digits, combinations);
        return combinations;
    }

    private static void backtrack(int index, StringBuilder currStr, String digits, List<String> combinations) {
        if(digits.length() == currStr.length()) {
            combinations.add(currStr.toString());
            return;
        }

        for (char ch : phoneMap.get(digits.charAt(index)).toCharArray()) {
            //choose
            currStr.append(ch);
            //explore
            backtrack(index+1, currStr, digits, combinations);
            //unchoose
            currStr.deleteCharAt(currStr.length()-1);
        }
    }
    public static void main(String[] args) {
        String digits1 = "24";
        String digits2 = "";
        System.out.println(letterCombinations(digits1));
        System.out.println(letterCombinations(digits2));

    }
}